tempesta_fw/sched/tfw_sched_hash.c

```diff
@@ -114,42 +229,73 @@ static int
 tfw_sched_hash_del_srv(TfwServer *srv)
 {
 	int ret = 0;
-	int i;
-
-	spin_lock_bh(&servers_write_lock);
-	i = find_srv_idx(srv);
-	if (i < 0) {
-		ERR("Can't delete the server from the scheduler - not found\n");
-		ret = -ENOENT;
-	} else {
-		servers[i] = servers[servers_n - 1];
-		--servers_n;
-		servers[servers_n] = NULL;
+	TfwSrvHash *curr;
+	TfwSrvHashList *new_list, *old_list;
+	size_t new_list_size, old_list_size;
+
+	spin_lock_bh(&srv_hashes_update_lock);
+
+	old_list = tfw_srv_hash_list;
+	old_list_size = TFW_SRV_HASH_LIST_SIZE(old_list->n);
+	new_list_size = TFW_SRV_HASH_LIST_SIZE(old_list->n - 1);
+
+	new_list = kzalloc(new_list_size, GFP_ATOMIC);
+	if (!new_list) {
+		ret = -ENOMEM;
+		goto out;
+	}
+
+	/* Copy entries to the new list excluding the one being deleted. */
+	for (curr = old_list->srv_hashes; curr->srv; ++curr) {
+		if (curr->srv != srv)
+			new_list->srv_hashes[new_list->n++] = *curr;
 	}
-	spin_unlock_bh(&servers_write_lock);
+	BUG_ON(new_list->n != (old_list->n - 1));
 
+	rcu_assign_pointer(tfw_srv_hash_list, new_list);
+	kfree_rcu(old_list, rcu);
+out:
+	spin_unlock_bh(&srv_hashes_update_lock);
 	return ret;
 }
 
+static TfwScheduler tfw_sched_hash_mod = {
+	.name = "hash",
+	.get_srv = tfw_sched_hash_get_srv,
+	.add_srv = tfw_sched_hash_add_srv,
+	.del_srv = tfw_sched_hash_del_srv
+};
+
 int
 tfw_sched_hash_init(void)
 {
-	static TfwScheduler tfw_sched_hash_mod = {
-		.name = "hash",
-		.get_srv = tfw_sched_hash_get_srv,
-		.add_srv = tfw_sched_hash_add_srv,
-		.del_srv = tfw_sched_hash_del_srv
-	};
+	int r;
 
 	LOG("init\n");
 
-	return tfw_sched_register(&tfw_sched_hash_mod);
+	/* Allocate a dummy list to avoid NULL checks in other functions. */
+	tfw_srv_hash_list = kzalloc(TFW_SRV_HASH_LIST_SIZE(0), GFP_KERNEL);
```

Comment:
Why to allocate an empty array if you know that you'll insert some entries soon? Why not to allocate some adequate values, like 16?